javaclient/src/main/java/confluo/streaming/ConfluoProducer.java

```diff
@@ -0,0 +1,165 @@
+    public void start() throws TException{
+        client= new RpcClient(host, port);
+            try {
+                long atomicLogId = client.getAtomicMultilog(topic);
+                client.setCurrentAtomicMultilog(topic);
+                logger.info(String.format("%s exist,atomic multilog id %d,reuse it now", topic, atomicLogId));
+            } catch (TApplicationException e) {
+                // process concurrently create topic
+                client.createAtomicMultilog(topic, produceSchema, storageMode);
+            } catch (TException  e){
+                logger.info("start error",e);
+                throw e;
+            }
+        curSchema=client.getSchema();
+        if(batchEnable) {
+            batchBuilder = new MessageBatchBuilder(client.getSchema());
+        }
+    }
+
+
+    /**
+     *
+     *  for a pub/sub system,it's schemaless,but here we define a relatively relaxed
+     *  schema{ long:a;msg:STRING(len)} with default timestamp,msg with fixed length is the actual payload for
+     *  producer
+     * @param  len message body length
+     * @return  schema string
+     **/
+     private String createSchema(int len){
+      return String.format("{msg: STRING(%d)}",len);
+     }
+
+     /**
+      * @return  current schema
+      *
+      **/
+     public Schema getSchema(){
+         return curSchema;
+     }
+
+
+    /**
+     *
+     * fill timestamp and send message
+     *
+     **/
+    public void send(ByteBuffer message) throws TException{
+        try {
+                message.putLong(0, System.currentTimeMillis());
+                message.clear();
+                if(batchEnable) {
+                    batchBuilder.addRecord(message);
+                    batched++;
+                    if(batched>=produceBatchSize){
+                        long offset=client.appendBatch(batchBuilder.getBatch());
+                        //logger.info("write offset:"+offset);
+                        batchBuilder.clear();
+                        totalProducedNum +=batched;
+                        batched=0;
+                    }
+                }else {
+                    client.appendRaw(message);
+                    totalProducedNum++;
+                }
+        }catch (IOException e){
+            logger.error("io error",e);
+        }
+    }
+
+    /**
+     * @return  get total produced messages
+     *
+     **/
+    public long getTotalProducedNum() {
+        return totalProducedNum;
+    }
+
+    /**
+     *
+     *  flush the last batch message
+     *
+     **/
+    public void flush() throws IOException,TException{
+        if(batchEnable&&batched>0){
+            long offset=client.appendBatch(batchBuilder.getBatch());
+            //logger.info("write offset:"+offset);
+            batchBuilder.clear();
+            totalProducedNum +=batched;
+            batched=0;
+        }
+    }
+
+    /**
+     * stop and close client
+     *
+     **/
+    public void stop() throws  TException{
```

Comment:
A cleaner approach would be to extend [Closeable](https://docs.oracle.com/javase/8/docs/api/index.html?java/io/Closeable.html), and implement this as close().

Reply:
ok